extensions/Lily/Data.js

```diff
@@ -0,0 +1,653 @@
+        if (list.value[i] === item1) {
+          newList.push(item2);
+        } else {
+          newList.push(list.value[i]);
+        }
+      }
+      list.value = newList;
+      list._monitorUpToDate = false;
+    }
+
+    repeatList(args, util) {
+      const list1 = getVarObjectFromName(args.LIST1, util, "list");
+      if (!list1) return;
+      const list2 = getVarObjectFromName(args.LIST2, util, "list");
+      if (!list2) return;
+      const currentVal = list1.value;
+      for (let i = 0; i < args.NUM; i++) {
+        list1.value = list1.value.concat(currentVal);
+      }
+    }
+
+    getListJoin(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return "";
+      return list.value.join(args.STRING);
+    }
+
+    timesItemAppears(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return 0;
+      return list.value.filter((model) => model == args.ITEM).length;
+    }
+
+    itemIndex(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return 0;
+      let indexes = [];
+      for (let index = 0; index < list.value.length; index++) {
+        if (list.value[index] === args.ITEM) {
+          indexes.push(index);
+        }
+      }
+      switch (args.INDEX) {
+        case "_first_":
+          return indexes[0] + 1;
+        case "_last_":
+          return indexes[indexes.length - 1] + 1;
+        case "_random_":
+          return indexes[Math.floor(Math.random() * indexes.length)] + 1;
+        default:
+          return indexes[args.INDEX - 1] + 1;
+      }
+    }
+
+    listIsEmpty(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return true;
+      if (list.value.length > 0) return false;
+      return true;
+    }
+
+    itemNumExists(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return false;
+      const listIndex = Scratch.Cast.toListIndex(
+        args.NUM,
+        list.value.length,
+        false
+      );
+      if (listIndex === Scratch.Cast.LIST_INVALID) return false;
+      return true;
+    }
+
+    orderIs(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return false;
+      const defaultList = cloneObj(list.value);
+      const orderedList = cloneObj(list.value).sort();
+      if (args.ORDER === "ascending") {
+        return arraysEqual(defaultList, orderedList);
+      }
+      if (args.ORDER === "descending") {
+        return arraysEqual(defaultList.reverse(), orderedList);
+      }
+      return false;
+    }
+
+    orderList(args, util) {
+      const list = getVarObjectFromName(args.LIST, util, "list");
+      if (!list) return;
+      if (args.ORDER === "reversed") {
+        list.value.reverse();
+      } else if (args.ORDER === "randomised") {
+        const randomised = list.value
```

Comment:
there are faster ways to do this https://www.google.com/search?hl=en&q=fisher%20yates%20random